### PR TITLE
Caliper

### PR DESCRIPTION
**Pull request by bslazarov**
The PR adds Caliper configuration files, simple macros and two instrumented examples - ex1 and ex1p. 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2104](https://github.com/mfem/mfem/pull/2104) | @bslazarov | @tzanio | @adrienbernede + @artv3 | 03/17/21 | 03/30/21 | 04/19/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by artv3 — APPROVED** [review]
This is cool: 

```
Path                       Time (E) Time (I) Time % (E) Time % (I) 
main                       0.025340 0.273509   9.264777 100.000000 
  Save the results         0.041577 0.041577  15.201328  15.201328 
  Solve A X=B              0.000037 0.168355   0.013528  61.553733 
    PCG                    0.168318 0.168318  61.540205  61.540205 
  Set up the bilinear form 0.025436 0.025436   9.299877   9.299877 
  Set up the linear form   0.012801 0.012801   4.680285   4.680285 
```

Questions:
1. What are the other caliper options supported? Should we list some examples? 

2. Is E and I exclusive and inclusive respectively? Can we choose one or the other?

**Comment by rcarson3** [comment]
> This is cool:
> 
> ```
> Path                       Time (E) Time (I) Time % (E) Time % (I) 
> main                       0.025340 0.273509   9.264777 100.000000 
>   Save the results         0.041577 0.041577  15.201328  15.201328 
>   Solve A X=B              0.000037 0.168355   0.013528  61.553733 
>     PCG                    0.168318 0.168318  61.540205  61.540205 
>   Set up the bilinear form 0.025436 0.025436   9.299877   9.299877 
>   Set up the linear form   0.012801 0.012801   4.680285   4.680285 
> ```
> 
> Questions:
> 
>     1. What are the other caliper options supported? Should we list some examples?
> 
>     2. Is E and I exclusive and inclusive respectively? Can we choose one or the other?

It should be inclusive and exclusive times. Caliper has quite a few different options.
I think my go to for ExaConstit when using a caliper config file make use of the following which I pulled off a caliper issue:
```
# [mpi-report]
CALI_SERVICES_ENABLE=aggregate,event,mpi,mpireport,timestamp
CALI_MPI_BLACKLIST=MPI_Comm_rank,MPI_Comm_size,MPI_Wtick,MPI_Wtime
CALI_EVENT_ENABLE_SNAPSHOT_INFO=true
CALI_TIMER_SNAPSHOT_DURATION=true
CALI_TIMER_INCLUSIVE_DURATION=true
CALI_MPIREPORT_FILENAME=stderr
CALI_MPIREPORT_CONFIG="select min(sum#time.inclusive.duration) as \"Min time/rank\", max(sum#time.inclusive.duration) as \"Max time/rank\", avg(sum#time.inclusive.duration) as \"Inc Avg time/rank\", percent_total(sum#time.duration) as \"Time % (total)\" group by prop:nested format tree"
```

**Comment by bslazarov** [comment]
One can supply any option supported by the Caliper Config Manager. The options depend also on how Caliper is configured. I would use the Caliper documentation rather than writing all possible options in the MFEM examples. 

 I added one additional example on how to run the code and I am referring the users to the Caliper documentation for more info on the topic.

**Comment by artv3** [comment]
> One can supply any option supported by the Caliper Config Manager. The options depend also on how Caliper is configured. I would use the Caliper documentation rather than writing all possible options in the MFEM examples.
> 
> I added one additional example on how to run the code and I am referring the users to the Caliper documentation for more info on the topic.
> 
> I am not sure which one of the internal MFEM methods should be instrumented. I have added instrumentation for the PCG solver only and I would like to hear suggestions and opinions on what should (or should not) be instrumented.

I think as a first step this is fine. Depending on the application, they may want or not want certain routines.... this is hard one for a library

**Comment by adrienbernede** [comment]
I'm really interested in this feature. I haven't looked at the documentation yet, nor have I reviewed the code, but I have 1 question:
Why creating MFEM overrides for CALI macros?

**Comment by bslazarov** [comment]
> I'm really interested in this feature. I haven't looked at the documentation yet, nor have I reviewed the code, but I have 1 question:
> Why creating MFEM overrides for CALI macros?

The reason is to avoid #ifdef statements and to make the code more readable. One can use the standard Caliper macros without any limitations.

**Comment by adrienbernede** [comment]
OK I see it now.

I guess I am just not found of the naming `MFEM_MARK_...` because is does not give a clue of what it does, except "marking", and is a bit misleading with the MFEM prefix.

What about `PERF_MARK_...` or similar?

**Comment by bslazarov** [comment]
> OK I see it now.
> 
> I guess I am just not found of the naming `MFEM_MARK_...` because is does not give a clue of what it does, except "marking", and is a bit misleading with the MFEM prefix.
> 
> What about `PERF_MARK_...` or similar?

I think MFEM should be there - so it is clear that this marking is coming from MFEM in care users employ MFEM in derived codes. I can do MFEM_PERF_MARK_.... It gets a bit longer - let me know what do you think.

File: makefile
Comment (adrienbernede) [inline makefile:335]: I think it would make more sense to group the `MFEM_USE_xxx` definitions:
```suggestion
 MFEM_USE_MKL_CPARDISO MFEM_USE_AMGX MFEM_USE_MUMPS MFEM_USE_CALIPER\
```

File: makefile
Comment (adrienbernede) [inline makefile:336]: I think it would make more sense to group the `MFEM_USE_xxx` definitions:
```suggestion
 MFEM_SOURCE_DIR MFEM_INSTALL_DIR
```

File: general/annotation.hpp
Comment (adrienbernede) [inline general/annotation.hpp:29]: I maintain that my preference goes for something like `MFEM_BENCHMARK_FUNCTION`. I would also keep it shorter with `MFEM_BENCHMARK_BEGIN` and `MFEM_BENCHMARK_END`.

File: examples/caliper/CMakeLists.txt
Comment (adrienbernede) [inline examples/caliper/CMakeLists.txt:19]: Indentation
```suggestion
  list(APPEND CALIPER_EXE_SRCS
```

File: examples/caliper/CMakeLists.txt
Comment (adrienbernede) [inline examples/caliper/CMakeLists.txt:30]: I would prefer the full word here
```suggestion
set(PREFIX caliper_)
add_mfem_examples(CALIPER_EXE_SRCS ${PREFIX})
```

File: examples/CMakeLists.txt
Comment (adrienbernede) [inline examples/CMakeLists.txt:151]: Creating `examples/caliper` leads to duplications, and will get worse with increasing use of Caliper.

File: examples/makefile
Comment (adrienbernede) [inline examples/makefile:58]: Creating `examples/caliper` leads to duplication, and will get worse with increasing use of Caliper.

**Review by adrienbernede — CHANGES_REQUESTED** [review]
I think we need to discuss the choice to create an additional directory for caliper cases in `examples`.
In fact, I think using Caliper could, and should eventually be extended to multiple examples. We should not duplicate the test case each time we want to have it run with benchmarking.

That's why I would suggest allowing Caliper to be used in any example, either as soon as `MFEM_USE_CALIPER` is defined, or triggered with specific options like `MFEM_BENCHMARK=ON` (although the latter seems redundant).

I would like to confirm that I would prefer `BENCH_MARK` rather than `MARK` in the annotations, to make it explicit that we are talking about performance annotations. Adding `REGION` when `BEGIN/END` is used seems redundant and could be skipped IMHO.

Thank you for introducing this tool to the MFEM project, I'm looking forward to introducing it as part of our testing process!

**Comment by bslazarov** [comment]
> I think we need to discuss the choice to create an additional directory for caliper cases in `examples`.
> In fact, I think using Caliper could, and should eventually be extended to multiple examples. We should not duplicate the test case each time we want to have it run with benchmarking.

I do agree that in the future we should use Caliper to benchmark the code. On the other hand we would like to have a simple introductory example. The additional Caliper statements might be confusing for the beginners. I am not sure what is the solution in this case and I am open for suggestions. If everyone agree that we should instrument the original example 1 - I can easily do it. 
For the other examples it is not that critical.

File: makefile
Comment (tzanio) [reply makefile:335]: Agreed

File: general/annotation.hpp
Comment (tzanio) [reply general/annotation.hpp:29]: My preference is for `MFEM_PERF_FUNCTION`, `MFEM_PERF_BEGIN` and `MFEM_PERF_END`

File: examples/caliper/CMakeLists.txt
Comment (tzanio) [reply examples/caliper/CMakeLists.txt:30]: Agreed

File: examples/CMakeLists.txt
Comment (tzanio) [reply examples/CMakeLists.txt:151]: Discussed at the weekly PR meeting

**Review by daboehme — APPROVED** [review]
Looks good to me! Seems most of the issues in the discussion above are MFEM style questions, and it's entirely up to you how to resolve them. It may be useful to add more annotations inside the MFEM library in the long run, but they can be added in the future as needed. This PR does all the major work of adding Caliper support in the first place.

**Comment by adrienbernede** [comment]
Tested locally, works for me. Waiting for the change in naming for approval, but it’s all good!

File: general/annotation.hpp
Comment (bslazarov) [reply general/annotation.hpp:29]: Set to:
MFEM_PERF_FUNCTION
MFEM_PERF_BEGIN(s)
MFEM_PERF_END(s)

File: makefile
Comment (bslazarov) [reply makefile:335]: Done

**Review by adrienbernede — APPROVED** [review]

**Comment by adrienbernede** [comment]
LGTM

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

File: CHANGELOG
Comment (v-dobrev) [inline CHANGELOG:29]:
```suggestion
- Added support for Caliper: a library to integrate performance profiling 
  capabilities into applications.
```

File: config/config.hpp.in
Comment (v-dobrev) [inline config/config.hpp.in:162]:
```suggestion
// Enable functionality based on the Caliper library.
// #define MFEM_USE_CALIPER

```

File: examples/caliper/ex1.cpp
Comment (v-dobrev) [inline examples/caliper/ex1.cpp:67]:
```suggestion
   // Caliper instrumentation
   MFEM_PERF_FUNCTION;

```

File: examples/caliper/ex1.cpp
Comment (v-dobrev) [inline examples/caliper/ex1.cpp:94]:
```suggestion
   args.AddOption(&cali_config, "-p", "--caliper",
                  "Caliper configuration string.");
```

File: examples/caliper/ex1.cpp
Comment (v-dobrev) [inline examples/caliper/ex1.cpp:109]:
```suggestion
   // Caliper configuration
```

File: examples/caliper/ex1p.cpp
Comment (v-dobrev) [inline examples/caliper/ex1p.cpp:68]:
```suggestion
   // Caliper instrumentation
```

File: examples/caliper/ex1p.cpp
Comment (v-dobrev) [inline examples/caliper/ex1p.cpp:117]:
```suggestion
   // Caliper configuration
```

File: examples/caliper/ex1.cpp
Comment (v-dobrev) [inline examples/caliper/ex1.cpp:199]: Similar to the parallel version, it probably makes sense to move this after `FormLinearSystem`.

File: examples/caliper/makefile
Comment (v-dobrev) [inline examples/caliper/makefile:60]:
```suggestion
ex1-test-seq: ex1
	@$(call mfem-test,$<,, Caliper serial example)
ex1p-test-par: ex1p
	@$(call mfem-test,$<, $(RUN_MPI), Caliper parallel example)
```

File: makefile
Comment (v-dobrev) [inline makefile:336]:
```suggestion
 MFEM_SOURCE_DIR MFEM_INSTALL_DIR
```

File: examples/caliper/ex1.cpp
Comment (bslazarov) [reply examples/caliper/ex1.cpp]: agree - corrected

File: INSTALL
Comment (v-dobrev) [inline INSTALL:727]: It may be good to add a version requirement for Caliper, similar to other packages. If you have not tested with older versions, just use the version you tested with and add "(may work with older versions too)" -- no need to check which older versions actually work.

Comment (bslazarov) [reply INSTALL:727]: It should work with older version, however, since it is tested with the current one 2.5.0 I have mentioned it in the INSTALL file

**Comment by tzanio** [comment]
Re-merged in `next`.

**Review by tzanio — APPROVED** [review]
Thank you so much @bslazarov for this really useful addition 👍 

_@v-dobrev, @artv3, @daboehme and @adrienbernede -- Thank you for reviewing!_

**Pull request merged**
